### PR TITLE
support for redhat simple signature verification

### PR DESCRIPTION
**Pull request by sjhx**
https://github.com/IBM/portieris/issues/70

### CODE REVIEW

File: helm/portieris/values.yaml [hunk 1]
```diff
@@ -8,7 +8,7 @@ namespace: ibm-system
 image:
   host: docker.io/ibmcom
   image: portieris
-  tag: 0.5.3
+  tag: simpla-alphae
```
[the remark this answers is not in the document]
Comment (sjhx) [reply helm/portieris/values.yaml]: done

File: helm/portieris/Chart.yaml [hunk 2]
```diff
@@ -1,8 +1,9 @@
 
 apiVersion: v1
 name: portieris
-version: 0.5.3
+version: 0.6.0-simple
```
[the remark this answers is not in the document]
Comment (sjhx) [reply helm/portieris/Chart.yaml]: done

File: pkg/controller/multi/controller.go [hunk 3]
```diff
@@ -96,120 +96,115 @@ func (c *Controller) mutatePodSpec(namespace, specPath string, pod corev1.PodSpe
...
-				a.StringToAdmissionResponse(err.Error())
-				continue containerLoop
-			} else if policy == nil || !(policy.Trust.Enabled != nil && *policy.Trust.Enabled == true) {
-				a.SetAllowed()
-				continue containerLoop
+			glog.Infof("Getting policy for container image: %s   namespace: %s", img.String(), namespace)
+			containerPolicy, err := c.policyClient.GetPolicyToEnforce(namespace, img.String())
+			if err != nil {
+				a.ToAdmissionResponse(err)
+				return a.Flush()
```
[the remark this answers is not in the document]
Comment (sjhx) [reply pkg/controller/multi/controller.go]: done

File: pkg/verifier/simple/imagePolicy.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/IBM/portieris/pkg/apis/securityenforcement/v1beta1"
+	"github.com/containers/image/v5/docker"
+	"github.com/containers/image/v5/image"
+	"github.com/containers/image/v5/manifest"
+	"github.com/containers/image/v5/signature"
+	"github.com/containers/image/v5/types"
+)
+
+// VerifyByPolicy1 ...
+func VerifyByPolicy1(imageToVerify string, credentials [][]string, portierisPolicy *v1beta1.Policy) (*bytes.Buffer, error, error) {
```
[the remark this answers is not in the document]
Comment (sjhx) [reply pkg/verifier/simple/imagePolicy.go]: done

File: pkg/verifier/simple/imagePolicy_test.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+// }
+
+// func TestVerifyByPolicy_GoodPathOther(t *testing.T) {
+
+// 	_, err := VerifyByPolicy("uk.icr.io/stuarts/other:latest", policySignedByStuart, username, password)
+// 	assert.NotNil(t, err, "expected error")
+// }
+
+func TestVerifyByPolicy(t *testing.T) {
+}
```
[the remark this answers is not in the document]
Comment (sjhx) [reply pkg/verifier/simple/imagePolicy_test.go]: was

**Comment by sjhx** [comment]
I think this represents a reasonable initial release of support for RedHat simple signatures with follow on issues to make incremental improvements, e.g. we should have a README to describe the policy options for simple.

File: README.md [hunk 6]
```diff
@@ -2,17 +2,20 @@
...
-Portieris is a Kubernetes admission controller for enforcing Content Trust. You can create image security policies for each Kubernetes namespace, or at the cluster level, and enforce different levels of trust for different images.
+Portieris is a Kubernetes admission controller for the enforcment of image security policies. You can create image security policies for each Kubernetes namespace, or at the cluster level, and enforce different rules for different images.
 
 ## How it works
 
-Portieris uses a Kubernetes Mutating Admission Webhook to modify your Kubernetes resources at the point of creation, to ensure that Kubernetes pulls the signed version. When configured to do so, it enforces trust pinning, and blocks the creation of resources that use untrusted images.
+Portieris uses a Kubernetes Mutating Admission Webhook to modify your Kubernetes resources at the point of creation, to ensure that Kubernetes runs only policy compliant images. When configured to do so, it can enforce Docker Content Trust with optional trust pinning, or can verify signatures created using RedHat's simple signing model and will prevent the creation of resources using untrused or unverified images.
```
Comment (molepigeon) [inline README.md]: typo: untrused

File: README.md [hunk 7]
```diff
@@ -2,17 +2,20 @@
...
+When you create or edit a workload, the Kubernetes API server sends a request to Portieris. The AdmissionRequest contains the content of your workload. For each image in your workload, Portieris finds a matching security policy.
+If trust enforcement is enabled in the policy, Portieris pulls signature information for your image from the corresponding Notary server and, if a signed version of the image exists, creates a JSON patch to edit the image name in the workload to the signed image by digest. If a signer is defined in the policy, Portieris additionally checks that the image is signed by the specified role, and verifies that the specified key was used to sign the image.
+If simple signing is specified by the policy, Portieris will verify the signature using using the public key and identity rules supplied in the policy and if verified similarly mutates the image name to a digest referencee to ensure that concurrent tag changes cannot influence the image being puuled.
+While it is possible to require both DCT and Simple signing, alternatives are *not* allowed, if both are required conflicting digests will cause the image to be denied.
```
Comment (molepigeon) [inline README.md]: Did you want to put each of these sentences on their own line? If so, you need to put another newline between them - at the moment this renders to one block of text as before.

File: README.md [hunk 8]
```diff
@@ -2,17 +2,20 @@
...
+When you create or edit a workload, the Kubernetes API server sends a request to Portieris. The AdmissionRequest contains the content of your workload. For each image in your workload, Portieris finds a matching security policy.
+If trust enforcement is enabled in the policy, Portieris pulls signature information for your image from the corresponding Notary server and, if a signed version of the image exists, creates a JSON patch to edit the image name in the workload to the signed image by digest. If a signer is defined in the policy, Portieris additionally checks that the image is signed by the specified role, and verifies that the specified key was used to sign the image.
+If simple signing is specified by the policy, Portieris will verify the signature using using the public key and identity rules supplied in the policy and if verified similarly mutates the image name to a digest referencee to ensure that concurrent tag changes cannot influence the image being puuled.
```
Comment (molepigeon) [inline README.md]: typos: referencee, puuled

File: README.md (hunk 7 above)
Comment (molepigeon) [inline README.md]: Perhaps:

> While it is possible to require both Notary and simple signing, the two methods must agree on the signed digest for the image. If the two methods return different signed digests, the image is denied.

File: pkg/controller/multi/controller.go [hunk 9]
```diff
@@ -131,6 +131,7 @@ func (c *Controller) mutatePodSpec(namespace, specPath string, pod corev1.PodSpe
 				// convert digest to patch
 				glog.Infof("Mutation #: %s %d  Image name: %s", containerType, containerIndex, img.String())
 				if strings.Contains(container.Image, img.String()) {
+					// TODO: seems odd to have both tag and digest
```
Comment (molepigeon) [inline pkg/controller/multi/controller.go]: TODO here

File: pkg/kubernetes/secret.go [hunk 10]
```diff
@@ -35,6 +35,22 @@ type RegistriesStruct map[string]struct {
...
+	if err != nil {
+		glog.Error("Error: ", err)
+		return nil, err
+	}
+	glog.Infof("Found secret %s", secretName)
+	// Obtain the key data
+	if key, ok := secret.Data["key"]; ok {
+		return key, nil
+	}
+	return nil, fmt.Errorf("Secret %q in %q does not contain a \"key\" attribte", secretName, namespace)
```
Comment (molepigeon) [inline pkg/kubernetes/secret.go]: typo: attribte

File: pkg/verifier/simple/imagePolicy_test.go [hunk 11]
```diff
@@ -18,41 +18,77 @@ package simple
...
+					},
+				},
+			}
+			digest, deny, err := VerifyByPolicy(tt.image, tt.credentials, policy)
+			if tt.wantErr {
+				assert.Error(t, err)
+				assert.Contains(t, err.Error(), tt.errMsg, "unexpected error")
+			} else {
+				assert.NoError(t, err)
+				//assert.Equal(t, tt.wantList, got)
```
Comment (molepigeon) [inline pkg/verifier/simple/imagePolicy_test.go]: Is this commented out intentionally?

File: pkg/controller/multi/controller.go (hunk 9 above)
Comment (sjhx) [reply pkg/controller/multi/controller.go]: it is a TODO and should be investigated, it is not a change from master, would an issue be better?

Comment (sjhx) [reply pkg/controller/multi/controller.go]: made an issue https://github.com/IBM/portieris/issues/90

File: pkg/verifier/simple/imagePolicy_test.go (hunk 11 above)
Comment (sjhx) [reply pkg/verifier/simple/imagePolicy_test.go]: deleted is better

**Pull request merged**